pkg/helpers/helpers.go

```diff
@@ -33,3 +33,21 @@ func IsObjectTaggedToDelete(obj client.Object) bool {
 	annotation, ok := annotations[DeleteTagAnnotation]
 	return ok && annotation == "true"
 }
+
+func MergeMapStringString(modified *bool, existing map[string]string, desired *map[string]string) {
+	if existing == nil {
+		existing = map[string]string{}
+	}
+
+	// for each existing key value set, e.g. labels
+	// check if it's present in desired. if not add it to desired.
+	// e.g. preserving user added labels.
+	for existingKey, existingValue := range existing {
+		for desiredKey, desiredValue := range *desired {
+			if _, exists := (existing)[desiredKey]; !exists || existingValue != desiredValue {
+				(*desired)[existingKey] = existingValue
+				*modified = true
+			}
+		}
+	}
+}
```

Comment:
I think this is not working as expected. For updates, we should be updating the existing map rather then the desired map and I think the inner `for` loop is not needed  🤔 

```suggestion
func MergeMapStringString(modified *bool, existing map[string]string, desired map[string]string) {
	if existing == nil {
		existing = map[string]string{}
	}

	// for each desired key value set, e.g. labels
	// check if it's present in existing. if not add it to existing.
	// e.g. preserving user added labels.
	for desiredKey, desiredValue := range desired {
		if existingValue, exists := existing[desiredKey]; !exists || desiredValue != existingValue {
			existing[desiredKey] = desiredValue
			*modified = true
		}
	}
}
```

Reply:
Thanks @KevFan the desired vs exists read backwards to me initially. I understood that desired is the object that is used to make the update but as we discussed I see now that it's existing that is the object that is used on update. 👍 cc @eguzki